qiskit/extensions/quantum_initialiser/initialiser.py

```diff
@@ -0,0 +1,411 @@
+
+       It inherits from CompositeGate in the same way that the Fredkin (cswap) gate does. Therefore self.data is the
+       list of gates (in order) that must be applied to implement this meta-gate.
+
+        name = instruction name string
+        param = list of complex amplitudes
+        arg = list of pairs (Register, index)
+        circ = QuantumCircuit or CompositeGate containing this gate
+        """
+
+    def __init__(self, name, param, arg, circ=None):
+        """Create new initialise composite gate."""
+
+        self.num_qubits = Initialiser._log2(len(param))
+
+        # Check if param is a power of 2
+        if self.num_qubits == 0:
+            raise QISKitError("Desired vector not a power of 2")
+
+        # Check if number of desired qubits agrees with available qubits
+        if len(arg) != self.num_qubits:
+            raise QISKitError("Number of complex amplitudes do not correspond to the number of qubits")
+
+        super(Initialiser, self).__init__(name, param, arg, circ)
+
+        # call to generate the circuit that takes the desired vector to zero
+        self.gates_to_uncompute()
+        # remove zero rotations and double cnots
+        self.optimize_gates()
+        # invert the circuit to create the desired vector from zero (assuming the qubits are in the zero state)
+        self.inverse()
+
+    def nth_qubit_from_least_sig_qubit(self, nth):
+        """Return the qubit that is nth away from the least significant qubit (LSB), so n=0, corresponds to the LSB """
+
+        # if LSB is first (as is the case with the IBM QE) and significance is in order:
+        return self.arg[nth]
+        # if MSB is first: return self.arg[self.num_qubits - 1 - n] # equivalent to self.arg[-(n+1)]
+        # to generalize any mapping could be placed here or even taken from the user
+
+    def reapply(self, circ):
+        """Reapply this gate to the corresponding qubits in circ."""
+
+        self._modifiers(circ.initialise(self.param, self.arg))
+
+    def gates_to_uncompute(self):
+        """call to populate the self.data list with gates that takes the desired vector to zero"""
+
+        # kick start the peeling loop
+        remaining_param = self.param
+
+        for i in range(self.num_qubits):
+            # work out which rotations must be done to disentangle the LSB qubit (we peel away one qubit at a time)
+            remaining_param, thetas, phis = Initialiser._rotations_to_disentangle(remaining_param)
+
+            # perform the required rotations to decouple the LSB qubit (so that it can be "factored" out, leaving a
+            # shorter amplitude vector to peel away)
+            self._attach(self._multiplex(RZGate, i, phis))
+            self._attach(self._multiplex(RYGate, i, thetas))
+
+    @staticmethod
+    def _log2(orig):
+        """Static internal method to calc the log (base 2) of the input (greater than 2). Return zero if the input is
+        not a positive power of 2. """
```

Comment:
If this function is only used twice: once during `__init__` for checking that `param` is valid, and later in `multiplex()`, would it make sense to just check that it is a multiple of two (`param % 2 == 0`) during `__init__`, and then just rely on `numpy.log2`?